File: experiments/native-launcher-python/native_launcher/sidecar.py
#!/usr/bin/env python3
"""Read-only sidecar that observes atomically renamed event files.

The sidecar enforces the runtime limits defined in ``limits.json`` so the
spool directory and in-memory summary stay bounded.
"""

import datetime
import json
import os
import signal
import stat
import sys
import time
from pathlib import Path
from typing import Any

from .limits import Limits, load_limits
from .schema import validate_file
from .utils import atomic_write

POLL_INTERVAL = 0.05


def _lifecycle_log(path: Path | None, label: str) -> None:
    if path is None:
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    now = datetime.datetime.now(datetime.timezone.utc).isoformat()
    with open(path, "a", encoding="utf-8") as fh:
        fh.write(f"{label} {now}\n")
        fh.flush()
        os.fsync(fh.fileno())


def _validate_event_json_size(data: bytes, limits: Limits) -> bool:
    if len(data) > limits.max_event_json_bytes:
        print("sidecar: event file exceeds max size; dropping", file=sys.stderr, flush=True)
        return False
    return True


class Sidecar:
    def __init__(self, runtime_dir: Path) -> None:
        self.runtime_dir = Path(runtime_dir)
        self.events_dir = self.runtime_dir / "events"
        self.summary_path = self.runtime_dir / "summary.json"
        self.schema_path = self.runtime_dir / "event.schema.json"
        limits_path = self.runtime_dir / "limits.json"
        limits_schema_path = self.runtime_dir / "limits.schema.json"
        self.limits = load_limits(limits_path, limits_schema_path, required=True)
        lifecycle_path = os.environ.get("GOAL_DEVIN_LIFECYCLE_LOG")
        self.lifecycle_path = Path(lifecycle_path) if lifecycle_path else None
        self.consumed: set[str] = set()
        self.consumed_order: list[str] = []
        self.total_events: int = 0
        self.tools: dict[str, int] = {}
        self.profiles: dict[str, int] = {}
        self.last_event: dict | None = None
        self.running = True

    def _validate_event(self, event: dict) -> bool:
        if not self.schema_path.exists():
            print("sidecar: event schema missing; rejecting event", file=sys.stderr, flush=True)
            return False
        try:
            errors = validate_file(event, self.schema_path)
        except (OSError, json.JSONDecodeError) as exc:
            print(f"sidecar: schema invalid; rejecting event: {exc}", file=sys.stderr, flush=True)
            return False
        if errors:
            print(f"sidecar: schema errors: {errors}", file=sys.stderr, flush=True)
            return False
        return True

    def _check_field_lengths(self, event: Any) -> bool:
        if not isinstance(event, dict):
            print("sidecar: event is not a JSON object; rejecting", file=sys.stderr, flush=True)
            return False
        limits = self.limits
        for key in ("event", "tool_name", "profile", "observed_at"):
            value = event.get(key)
            if isinstance(value, str) and len(value) > limits.max_event_value_length:
                print(
                    f"sidecar: event {key} exceeds max length; dropping",
                    file=sys.stderr,
                    flush=True,
                )
                return False
        if (
            isinstance(event.get("tool_name"), str)
            and len(event["tool_name"]) > limits.max_tool_name_length
        ):
            return False
        if (
            isinstance(event.get("profile"), str)
            and len(event["profile"]) > limits.max_profile_length
        ):
            return False
        return True

    def _update_summary(self) -> None:
        limits = self.limits
        recent_ids = self.consumed_order[-limits.max_recent_event_ids :]
        summary: dict[str, Any] = {
            "schema_version": 1,
            "run_id": self.runtime_dir.name,
            "total_events": self.total_events,
            "consumed_event_ids": recent_ids,
            "tools": self.tools,
            "profiles": self.profiles,
            "last_event": self.last_event,
        }

        # Iteratively trim bounded fields until the serialized summary fits inside
        # max_summary_size_bytes.  The limits file is required to allow a minimal
        # summary, but this loop is the final fail-open guard against any path.
        def _serialize() -> str:
            return json.dumps(summary, indent=2)

        def _size(text: str) -> int:
            return len(text.encode("utf-8"))

        text = _serialize()
        if _size(text) > limits.max_summary_size_bytes:
            summary["consumed_event_ids"] = []
            text = _serialize()
        if _size(text) > limits.max_summary_size_bytes:
            summary["tools"] = {}
            text = _serialize()
        if _size(text) > limits.max_summary_size_bytes:
            summary["profiles"] = {}
            text = _serialize()
        if _size(text) > limits.max_summary_size_bytes:
            summary["last_event"] = None
            text = _serialize()
        if _size(text) > limits.max_summary_size_bytes:
            # Fallback to compact JSON for the smallest possible valid shape.
            text = json.dumps(summary, separators=(",", ":"))
            if _size(text) > limits.max_summary_size_bytes:
                raise RuntimeError(
                    f"summary cannot fit within max_summary_size_bytes "
                    f"({limits.max_summary_size_bytes})"
                )
        atomic_write(self.summary_path, text)

    def _add_tool(self, tool: str | None) -> None:
        if not isinstance(tool, str):
            return
        if tool == "run_subagent":
            if tool not in self.tools and len(self.tools) >= self.limits.max_distinct_tools:
                # Reserve a slot for the canonical run_subagent observation.
                for existing in list(self.tools.keys()):
                    if existing != "run_subagent":
                        del self.tools[existing]
                        break
        elif tool not in self.tools and len(self.tools) >= self.limits.max_distinct_tools:
            # Already at the distinct-tool limit; drop new tools fail-open.
            return
        self.tools[tool] = self.tools.get(tool, 0) + 1

    def _add_profile(self, profile: str | None) -> None:
        if not isinstance(profile, str):
            return
        if profile not in self.profiles and len(self.profiles) >= self.limits.max_distinct_profiles:
            return
        self.profiles[profile] = self.profiles.get(profile, 0) + 1

    def _reject_file(self, path: Path, event_id: str, reason: str) -> None:
        """Mark a spool file as terminally rejected and remove it if possible.

        Files that cannot be deleted are left in the spool but tracked in
        ``consumed`` so they are not reprocessed forever.  The trimmer will
        count them toward the total byte bound and remove them if necessary.
        """
        print(f"sidecar: rejecting {path.name}: {reason}", file=sys.stderr, flush=True)
        self.consumed.add(event_id)
        try:
            path.unlink()
            self.consumed.discard(event_id)
        except OSError:
            pass

    def _process_file(self, path: Path) -> None:
        if path.suffix != ".json":
            return
        event_id = path.stem
        if event_id in self.consumed:
            return
        try:
            data = path.read_bytes()
        except OSError as exc:
            self._reject_file(path, event_id, f"cannot read file: {exc}")
            return

        if not _validate_event_json_size(data, self.limits):
            self._reject_file(path, event_id, "exceeds max event JSON size")
            return

        try:
            text = data.decode("utf-8")
            event = json.loads(text)
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            self._reject_file(path, event_id, f"not valid UTF-8 JSON: {exc}")
            return
        except ValueError as exc:
            self._reject_file(path, event_id, f"JSON parse error: {exc}")
            return

        if not self._check_field_lengths(event):
            self._reject_file(path, event_id, "field length limit exceeded")
            return
        if not self._validate_event(event):
            self._reject_file(path, event_id, "schema validation failed")
            return

        self.consumed.add(event_id)
        self.consumed_order.append(event_id)
        self.total_events += 1
        tool = event.get("tool_name")
        profile = event.get("profile")
        self._add_tool(tool)
        if tool == "run_subagent" and isinstance(profile, str):
            if (
                profile not in self.profiles
                and len(self.profiles) >= self.limits.max_distinct_profiles
            ):
                # Reserve a slot for the worker profile observed through run_subagent.
                for existing in list(self.profiles.keys()):
                    if existing != profile:
                        del self.profiles[existing]
                        break
        self._add_profile(profile)
        self.last_event = {
            "tool_name": tool,
            "profile": profile,
            "is_background": event.get("is_background"),
            "success": event.get("success"),
            "observed_at": event.get("observed_at"),
        }
        self._update_summary()

    def _trim_spool(self) -> None:
        """Remove oldest consumed event files to keep count and size bounded."""
        limits = self.limits
        try:
            paths = list(self.events_dir.iterdir())
        except OSError:
            return

        entries: list[tuple[Path, int, float]] = []
        for p in paths:
            if p.suffix != ".json":
                continue
            try:
                st = p.lstat()
                if not stat.S_ISREG(st.st_mode):
                    continue
                entries.append((p, st.st_size, st.st_mtime))
            except OSError:
                continue

        if not entries:
            return

        entries.sort(key=lambda x: x[2])
        total = sum(size for _, size, _ in entries)
        while (
            len(entries) > limits.max_retained_event_files or total > limits.max_total_spool_bytes
        ):
            removed = False
            for i, (p, size, _) in enumerate(entries):
                if p.stem in self.consumed and p.exists():
                    try:
                        p.unlink()
                        self.consumed.discard(p.stem)
                        # Also discard from the ordered recent list, but keep
                        # total_events as an accurate observed count.
                        if p.stem in self.consumed_order:
                            self.consumed_order.remove(p.stem)
                        total -= size
                        entries.pop(i)
                        removed = True
                        break
                    except OSError:
                        pass
            if not removed:
                break

    def _drain(self) -> None:
        try:
            for name in os.listdir(self.events_dir):
                path = self.events_dir / name
                if path.is_file():
                    self._process_file(path)
        except OSError:
            pass
        self._trim_spool()

    def _write_pid(self) -> None:
        pid_path = self.runtime_dir / "sidecar.pid"
        try:
            with open(pid_path, "w", encoding="utf-8") as fh:
                fh.write(f"{os.getpid()}\n")
                fh.flush()
                os.fsync(fh.fileno())
            os.chmod(pid_path, 0o600)
        except OSError:
            pass

    def _signal_ready(self) -> None:
        ready_path = self.runtime_dir / "sidecar-ready"
        try:
            with open(ready_path, "w", encoding="utf-8") as fh:
                fh.write("ready\n")
                fh.flush()
                os.fsync(fh.fileno())
            os.chmod(ready_path, 0o600)
        except OSError:
            pass

    def run(self) -> int:
        self.events_dir.mkdir(parents=True, exist_ok=True)
        self._update_summary()
        self._write_pid()
        _lifecycle_log(self.lifecycle_path, "sidecar_start")
        self._signal_ready()

        def _stop(signum, _frame):  # noqa: ANN001
            self.running = False

        signal.signal(signal.SIGTERM, _stop)
        signal.signal(signal.SIGINT, _stop)

        while self.running:
            self._drain()
            time.sleep(POLL_INTERVAL)

        self._drain()
        self._update_summary()
        _lifecycle_log(self.lifecycle_path, "sidecar_stop")
        return 0


def main(argv: list[str] | None = None) -> int:
    argv = argv or sys.argv[1:]
    if len(argv) < 1:
        print("usage: sidecar.py <runtime-dir>", file=sys.stderr, flush=True)
        return 2
    runtime_dir = Path(argv[0])
    return Sidecar(runtime_dir).run()


if __name__ == "__main__":
    sys.exit(main())

File: experiments/native-launcher-rust/src/sidecar.rs
use crate::hook::sanitize_string;
use crate::limits::{load_limits, Limits};
use crate::schema::{load_schema, validate};
use crate::utils::{atomic_write, lifecycle_log, utcnow_iso};
use anyhow::{anyhow, Result};
use serde_json::{Map, Value};
use std::collections::{HashMap, HashSet};
use std::fs;
use std::io::{self, Write};
use std::os::unix::fs::MetadataExt;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

const POLL_INTERVAL_MS: u64 = 50;

pub struct Sidecar {
    runtime_dir: PathBuf,
    events_dir: PathBuf,
    summary_path: PathBuf,
    schema_path: PathBuf,
    limits: Limits,
    lifecycle_path: Option<PathBuf>,
    consumed: HashSet<String>,
    consumed_order: Vec<String>,
    total_events: u64,
    tools: HashMap<String, u64>,
    profiles: HashMap<String, u64>,
    last_event: Option<Value>,
    profile_id: Option<String>,
    canonical_event_id: Option<String>,
    running: Arc<AtomicBool>,
}

impl Sidecar {
    pub fn new(runtime_dir: PathBuf) -> Result<Self> {
        let events_dir = runtime_dir.join("events");
        let summary_path = runtime_dir.join("summary.json");
        let schema_path = runtime_dir.join("event.schema.json");
        let limits_path = runtime_dir.join("limits.json");
        let limits_schema_path = runtime_dir.join("limits.schema.json");
        let limits = load_limits(Some(&limits_path), Some(&limits_schema_path), true)?;

        let manifest_path = runtime_dir.join("manifest.json");
        let profile_id = if manifest_path.exists() {
            crate::utils::read_json(&manifest_path).ok().and_then(|v| {
                v.get("profile_id")
                    .and_then(|p| p.as_str())
                    .map(|s| s.to_string())
            })
        } else {
            None
        };

        let lifecycle_path = std::env::var_os("GOAL_DEVIN_LIFECYCLE_LOG").map(PathBuf::from);

        Ok(Self {
            runtime_dir,
            events_dir,
            summary_path,
            schema_path,
            limits,
            lifecycle_path,
            consumed: HashSet::new(),
            consumed_order: Vec::new(),
            total_events: 0,
            tools: HashMap::new(),
            profiles: HashMap::new(),
            last_event: None,
            profile_id,
            canonical_event_id: None,
            running: Arc::new(AtomicBool::new(false)),
        })
    }

    fn _lifecycle(&self, label: &str) {
        if let Some(ref p) = self.lifecycle_path {
            let _ = lifecycle_log(p, label);
        }
    }

    fn _write_pid(&self) {
        let pid_path = self.runtime_dir.join("sidecar.pid");
        let data = format!("{}\n", std::process::id());
        let _ = atomic_write(&pid_path, &data, 0o600);
    }

    fn _signal_ready(&self) {
        let ready_path = self.runtime_dir.join("sidecar-ready");
        let _ = atomic_write(&ready_path, "ready\n", 0o600);
    }

    fn _validate_event(&self, event: &Value) -> Vec<String> {
        if !self.schema_path.exists() {
            return vec!["event schema missing".to_string()];
        }
        let schema = match load_schema(&self.schema_path) {
            Ok(s) => s,
            Err(e) => return vec![format!("cannot load event schema: {}", e)],
        };
        validate(event, &schema)
    }

    fn _check_field_lengths(&self, event: &Value) -> bool {
        let limits = &self.limits;
        for key in ["event", "tool_name", "profile", "observed_at"] {
            if let Some(s) = event.get(key).and_then(|v| v.as_str()) {
                if s.len() > limits.max_event_value_length {
                    return false;
                }
            }
        }
        if let Some(s) = event.get("tool_name").and_then(|v| v.as_str()) {
            if s.len() > limits.max_tool_name_length {
                return false;
            }
        }
        if let Some(s) = event.get("profile").and_then(|v| v.as_str()) {
            if s.len() > limits.max_profile_length {
                return false;
            }
        }
        true
    }

    fn _add_tool(&mut self, tool: Option<&str>) {
        let limits = &self.limits;
        if let Some(tool) = tool {
            if tool == "run_subagent" {
                if !self.tools.contains_key(tool) && self.tools.len() >= limits.max_distinct_tools {
                    // Reserve a slot
                    let to_remove: Option<String> =
                        self.tools.keys().find(|k| *k != "run_subagent").cloned();
                    if let Some(k) = to_remove {
                        self.tools.remove(&k);
                    }
                }
            } else if !self.tools.contains_key(tool)
                && self.tools.len() >= limits.max_distinct_tools
            {
                return;
            }
            *self.tools.entry(tool.to_string()).or_insert(0) += 1;
        }
    }

    fn _add_profile(&mut self, profile: Option<&str>, tool: Option<&str>) {
        let limits = &self.limits;
        if let Some(profile) = profile {
            let is_worker =
                self.profile_id.as_deref() == Some(profile) && tool == Some("run_subagent");
            if !self.profiles.contains_key(profile)
                && self.profiles.len() >= limits.max_distinct_profiles
            {
                if is_worker {
                    let to_remove: Option<String> =
                        self.profiles.keys().find(|k| *k != profile).cloned();
                    if let Some(k) = to_remove {
                        self.profiles.remove(&k);
                    }
                } else {
                    return;
                }
            }
            *self.profiles.entry(profile.to_string()).or_insert(0) += 1;
        }
    }

    fn _update_summary(&self) -> Result<()> {
        let recent: Vec<Value> = self
            .consumed_order
            .iter()
            .rev()
            .take(self.limits.max_recent_event_ids)
            .rev()
            .map(|s| Value::String(s.clone()))
            .collect();

        let tools: Map<String, Value> = self
            .tools
            .iter()
            .map(|(k, v)| (k.clone(), Value::Number((*v).into())))
            .collect();
        let profiles: Map<String, Value> = self
            .profiles
            .iter()
            .map(|(k, v)| (k.clone(), Value::Number((*v).into())))
            .collect();

        let mut summary = Map::new();
        summary.insert("schema_version".to_string(), Value::Number(1.into()));
        summary.insert(
            "run_id".to_string(),
            Value::String(
                self.runtime_dir
                    .file_name()
                    .unwrap_or_default()
                    .to_string_lossy()
                    .into_owned(),
            ),
        );
        summary.insert(
            "total_events".to_string(),
            Value::Number(self.total_events.into()),
        );
        summary.insert("consumed_event_ids".to_string(), Value::Array(recent));
        summary.insert("tools".to_string(), Value::Object(tools));
        summary.insert("profiles".to_string(), Value::Object(profiles));
        summary.insert(
            "last_event".to_string(),
            self.last_event.clone().unwrap_or(Value::Null),
        );

        let mut data = serde_json::to_string_pretty(&Value::Object(summary.clone()))?;
        if data.as_bytes().len() > self.limits.max_summary_size_bytes {
            summary.insert("consumed_event_ids".to_string(), Value::Array(vec![]));
            data = serde_json::to_string_pretty(&Value::Object(summary.clone()))?;
        }
        if data.as_bytes().len() > self.limits.max_summary_size_bytes {
            summary.insert("tools".to_string(), Value::Object(Map::new()));
            data = serde_json::to_string_pretty(&Value::Object(summary.clone()))?;
        }
        if data.as_bytes().len() > self.limits.max_summary_size_bytes {
            summary.insert("profiles".to_string(), Value::Object(Map::new()));
            data = serde_json::to_string_pretty(&Value::Object(summary.clone()))?;
        }
        if data.as_bytes().len() > self.limits.max_summary_size_bytes {
            summary.insert("last_event".to_string(), Value::Null);
            data = serde_json::to_string_pretty(&Value::Object(summary.clone()))?;
        }
        if data.as_bytes().len() > self.limits.max_summary_size_bytes {
            data = serde_json::to_string(&Value::Object(summary.clone()))?;
        }
        if data.as_bytes().len() > self.limits.max_summary_size_bytes {
            return Err(anyhow!("summary cannot fit within max_summary_size_bytes"));
        }

        atomic_write(&self.summary_path, &data, 0o600)?;
        Ok(())
    }

    fn _process_file(&mut self, path: &Path) {
        let event_id = match path.file_stem().and_then(|s| s.to_str()) {
            Some(s) => s.to_string(),
            None => return,
        };
        if path.extension().and_then(|s| s.to_str()) != Some("json") {
            return;
        }
        if self.consumed.contains(&event_id) {
            return;
        }

        let data = match fs::read(path) {
            Ok(d) => d,
            Err(_) => return,
        };

        if data.len() > self.limits.max_event_json_bytes {
            let _ = fs::remove_file(path);
            return;
        }

        let event: Value = match serde_json::from_slice(&data) {
            Ok(v) => v,
            Err(_) => {
                let _ = fs::remove_file(path);
                return;
            }
        };

        if !self._check_field_lengths(&event) {
            let _ = fs::remove_file(path);
            return;
        }

        let errors = self._validate_event(&event);
        if !errors.is_empty() {
            let _ = fs::remove_file(path);
            return;
        }

        self.consumed.insert(event_id.clone());
        self.consumed_order.push(event_id.clone());
        self.total_events += 1;

        let tool = event
            .get("tool_name")
            .and_then(|v| v.as_str())
            .map(|s| sanitize_string(s, self.limits.max_tool_name_length));
        let profile = event
            .get("profile")
            .and_then(|v| v.as_str())
            .map(|s| sanitize_string(s, self.limits.max_profile_length));

        self._add_tool(tool.as_deref());
        self._add_profile(profile.as_deref(), tool.as_deref());

        if event.get("event").and_then(|v| v.as_str()) == Some("PostToolUse")
            && event.get("success").and_then(|v| v.as_bool()) == Some(true)
            && tool.as_deref() == Some("run_subagent")
            && profile.as_deref() == self.profile_id.as_deref()
        {
            self.canonical_event_id = Some(event_id.clone());
        }

        let mut last = Map::new();
        last.insert(
            "tool_name".to_string(),
            tool.map(Value::String).unwrap_or(Value::Null),
        );
        last.insert(
            "profile".to_string(),
            profile.map(Value::String).unwrap_or(Value::Null),
        );
        last.insert(
            "is_background".to_string(),
            event.get("is_background").cloned().unwrap_or(Value::Null),
        );
        last.insert(
            "success".to_string(),
            event.get("success").cloned().unwrap_or(Value::Null),
        );
        last.insert(
            "observed_at".to_string(),
            event
                .get("observed_at")
                .and_then(|v| v.as_str())
                .map(|s| sanitize_string(s, self.limits.max_event_value_length))
                .map(Value::String)
                .unwrap_or_else(|| Value::String(utcnow_iso())),
        );
        self.last_event = Some(Value::Object(last));

        let _ = self._update_summary();
    }

    fn _trim_spool(&mut self) {
        let mut entries: Vec<(PathBuf, u64, u64)> = Vec::new();
        if let Ok(iter) = fs::read_dir(&self.events_dir) {
            for entry in iter.flatten() {
                let p = entry.path();
                if p.extension().and_then(|s| s.to_str()) != Some("json") {
                    continue;
                }
                if let Ok(meta) = entry.metadata() {
                    if !meta.is_file() {
                        continue;
                    }
                    let mtime = meta.mtime() as u64;
                    let size = meta.size();
                    entries.push((p, size, mtime));
                }
            }
        }
        if entries.is_empty() {
            return;
        }

        entries.sort_by(|a, b| a.2.cmp(&b.2));
        let mut total: u64 = entries.iter().map(|e| e.1).sum();

        while entries.len() > self.limits.max_retained_event_files
            || total > self.limits.max_total_spool_bytes as u64
        {
            let mut removed = false;
            for i in 0..entries.len() {
                let (p, size, _) = &entries[i];
                let stem = p.file_stem().and_then(|s| s.to_str()).unwrap_or("");
                if self.canonical_event_id.as_deref() == Some(stem) {
                    continue;
                }
                if self.consumed.contains(stem) && p.exists() && fs::remove_file(p).is_ok() {
                    self.consumed.remove(stem);
                    self.consumed_order.retain(|s| s != stem);
                    total -= size;
                    entries.remove(i);
                    removed = true;
                    break;
                }
            }
            if !removed {
                break;
            }
        }
    }

    fn _drain(&mut self) {
        if let Ok(iter) = fs::read_dir(&self.events_dir) {
            for entry in iter.flatten() {
                let p = entry.path();
                if p.is_file() {
                    self._process_file(&p);
                }
            }
        }
        self._trim_spool();
    }

    pub fn run(&mut self) -> i32 {
        if fs::create_dir_all(&self.events_dir).is_err() {
            return 1;
        }
        let _ = self._update_summary();
        self._write_pid();
        self._lifecycle("sidecar_start");
        self._signal_ready();

        let running = Arc::clone(&self.running);
        let term_id =
            signal_hook::flag::register(signal_hook::consts::SIGTERM, Arc::clone(&running));
        if let Err(e) = &term_id {
            self._lifecycle(&format!("sigterm_register_failed:{}", e));
        }
        let int_id = signal_hook::flag::register(signal_hook::consts::SIGINT, Arc::clone(&running));
        if let Err(e) = &int_id {
            self._lifecycle(&format!("sigint_register_failed:{}", e));
        }

        while !running.load(Ordering::Relaxed) {
            self._drain();
            std::thread::sleep(Duration::from_millis(POLL_INTERVAL_MS));
        }

        self._drain();
        let _ = self._update_summary();
        self._lifecycle("sidecar_stop");
        0
    }
}

pub fn run(runtime_dir: PathBuf) -> i32 {
    match Sidecar::new(runtime_dir) {
        Ok(mut sidecar) => sidecar.run(),
        Err(e) => {
            let _ = writeln!(io::stderr(), "sidecar: cannot initialize: {}", e);
            1
        }
    }
}
